src/messaging/discord.ts

```diff
@@ -122,47 +122,31 @@ export async function getDMResponseAsync(
   botMessage: Discord.Message | undefined,
   timeout: number
 ): Promise<string> {
-  const iterations = Math.max(Math.floor(timeout / pollInterval), 1);
-  let iteration = 0;
   const client = await getDiscordClientAsync();
   const dmChannel = await getDMChannelAsync(client);
   if (!dmChannel) {
     logger.error('unable to get discord DM channel');
     return '';
   }
-  return new Promise(resolve => {
-    let response = '';
-    const intervalId = setInterval(async () => {
-      const finish = (result: string) => {
-        clearInterval(intervalId);
-        resolve(result);
-      };
-      try {
-        iteration++;
-        const messages = await dmChannel.messages.fetch({
-          after: botMessage?.id,
-        });
-        const lastUserMessage = messages
-          .filter(message => message.reference?.messageID === botMessage?.id)
-          .last();
-        if (!lastUserMessage) {
-          if (iteration >= iterations) {
-            await dmChannel.send('Timed out waiting for response... 😿');
-            logger.error('✖ no response from user');
-            return finish(response);
-          }
-        } else {
-          response = lastUserMessage.cleanContent;
-          lastUserMessage.react('✅');
-          logger.info(`✔ got captcha response: ${response}`);
-          return finish(response);
-        }
-      } catch (error: unknown) {
-        logger.error("✖ couldn't get captcha response", error);
-        return finish(response);
-      }
-    }, pollInterval * 1000);
-  });
+  let response = '';
+  try {
+    const filter = (message: Discord.Message) =>
+      !message.reference || message.reference?.messageID === botMessage?.id;
```

Comment:
The documentation certainly needs to be updated. It was going to be my next job now that I'm mostly there with how to set things up. It didn't have any information about setting up an app and then bot in Discord to make this work. I'm also new to Discord (only got on because of this GPU shortage), so I wasn't even aware you could reply directly to a message until I started digging into why this was not working correctly.

So what about if I make it an option? The default being that a user can just "reply" in the DM channel like normal and then if they want to have multiple bots running, then have an option to change it to require them to reply directly to the individual message?

That'd make things easier for newbs like me in initial setup and then for those that want to go crazy and have multiple bots running, it's an easy change to make it work there too. The only downside with that would be with people that already have this set up with multiple bots. But since this whole notification thing for captchas is so new, that hopefully wouldn't be _too_ much of an issue.

Finally, for Slack, I really have no interest in creating an account there and learning their stuff. But if you really need the help to get that done, I probably can. I was going to next work on the issues with CloudFlare and get the bot a little more stealthy with the [puppeteer-extra-plugin-stealth](https://www.npmjs.com/package/puppeteer-extra-plugin-stealth) package next though...

Reply:
> The documentation certainly needs to be updated. It was going to be my next job now that I'm mostly there with how to set things up. It didn't have any information about setting up an app and then bot in Discord to make this work. I'm also new to Discord (only got on because of this GPU shortage), so I wasn't even aware you could reply directly to a message until I started digging into why this was not working correctly.

If I'm understanding correctly overall, your difficulties were with setting up the Discord bot and understanding that you had to reply to individual messages. I believe this is all in the ["How to use" and FAQ sections of the captcha reference](https://www.jef.buzz/streetmerchant/reference/captcha/#how-to-use), though I know the documentation regarding the latter issue wasn't in there until a few days ago. Excited to what updates would help draw your attention next time.

> So what about if I make it an option? The default being that a user can just "reply" in the DM channel like normal and then if they want to have multiple bots running, then have an option to change it to require them to reply directly to the individual message?
>
> That'd make things easier for newbs like me in initial setup and then for those that want to go crazy and have multiple bots running, it's an easy change to make it work there too. The only downside with that would be with people that already have this set up with multiple bots. But since this whole notification thing for captchas is so new, that hopefully wouldn't be _too_ much of an issue.

That sounds like a step back from what you've got already, tbh. Just to affirm, I don't have a problem with what you've got, just proposing the idea of adding helpful text to the DMs so users know what they can do. What your change is doing is ultimately solving user error, which is fine, and adding some guidance text can be another layer to help mitigate that as well. It could be something as simple as pointing to the usage instructions in the docs (though that could also break unfurling in Discord until the screenshot capture changes get merged). Anyway, just a suggestion, nothing else.

> Finally, for Slack, I really have no interest in creating an account there and learning their stuff. But if you really need the help to get that done, I probably can. 

I was just giving you the opportunity to add that change here while you're already working on it. It's totally fine if you can't do it!

> I was going to next work on the issues with CloudFlare 

Nice!

> and get the bot a little more stealthy with the [puppeteer-extra-plugin-stealth](https://www.npmjs.com/package/puppeteer-extra-plugin-stealth) package next though...

The stealth plugin was used in the past and removed. I can't provide any information as to why as I wasn't involved at that time (perhaps @jef can share), but I can already attest to the fact that adding it back does not fix any of the high-level issues people have faced lately with bot detection/browser mismatch. You *will* see it perform better on bot detection test sites, but things like using proxies on certain sites will still get flagged/trigger captcha.

Reply:
I'd certainly like to hear from @jef as to why it was removed. As things are now, I can't get [zotac](https://www.zotecstore.com/) to proceed past the CloudFlare check. My hope was that the plugin would help with that, so we'll see.

As for the proxies, there's nothing to be done about that in software. If an IP block is marked as being the source of bots, the only thing to do is use another IP.

It's always the problem with this kind of thing. Each side makes a change and then the other side works to overcome that new hurdle/trick.

Reply:
Sure, I don't use zotac so I wasn't planning on it, but if you really need help, I can run it through my local branch that has stealth integrated into main already.

The proxies problem has nothing to do with IP blocks, it has to do with how requests are generated through proxies. It sounds like it's not related to a problem you're trying to solve though.

Reply:
I removed stealth as types were brought in with Puppeteer 8 and didn't work with some of the plugins we had. So I effectively removed them. Some of the plugins removed were verbose and unneeded, but if stealth is need is some of these use cases, we can use it.